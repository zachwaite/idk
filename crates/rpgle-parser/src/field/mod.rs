mod comment_field;
mod control_level_field;
mod datastructure_type_field;
mod datatype_field;
mod decimals_field;
mod definition_type_field;
mod device_field;
mod endfile_field;
mod external_description_field;
mod factor1_field;
mod factor2_field;
mod file_addition_field;
mod file_designation_field;
mod file_format_field;
mod file_organization_field;
mod file_sequence_field;
mod filetype_field;
mod formtype_field;
mod idk_field;
mod indicators_field;
mod keylength_field;
mod keywords_field;
mod limits_processing_field;
mod name_field;
mod nothing_field;
mod operation_field;
mod position_field;
mod record_address_type_field;
mod record_length_field;
mod reserved_field;
mod result;
mod result_field;
mod result_length_field;
mod sequence_field;

pub use comment_field::CommentField;
pub use control_level_field::{ControlLevel, ControlLevelField};
pub use datastructure_type_field::{DatastructureType, DatastructureTypeField};
pub use datatype_field::{Datatype, DatatypeField};
pub use decimals_field::{Decimals, DecimalsField};
pub use definition_type_field::{DefinitionType, DefinitionTypeField};
pub use device_field::{Device, DeviceField};
pub use endfile_field::{Endfile, EndfileField};
pub use external_description_field::{ExternalDescription, ExternalDescriptionField};
pub use factor1_field::Factor1Field;
pub use factor2_field::Factor2Field;
pub use file_addition_field::{FileAddition, FileAdditionField};
pub use file_designation_field::{FileDesignation, FileDesignationField};
pub use file_format_field::{FileFormat, FileFormatField};
pub use file_organization_field::{FileOrganization, FileOrganizationField};
pub use file_sequence_field::{FileSequence, FileSequenceField};
pub use filetype_field::FiletypeField;
pub use formtype_field::{Formtype, FormtypeField};
pub use idk_field::IdkField;
pub use indicators_field::IndicatorsField;
pub use keylength_field::{KeyLength, KeyLengthField};
pub use keywords_field::KeywordsField;
pub use limits_processing_field::{LimitsProcessing, LimitsProcessingField};
pub use name_field::NameField;
pub use nothing_field::NothingField;
pub use operation_field::OperationField;
pub use position_field::{POSField, POS};
pub use record_address_type_field::{RecordAddressType, RecordAddressTypeField};
pub use record_length_field::RecordLengthField;
pub use reserved_field::{Reserved, ReservedField};
pub use result::FieldResult; // used by each field type
pub use result_field::ResultField;
pub use result_length_field::{ResultLength, ResultLengthField};
pub use sequence_field::SequenceField;
