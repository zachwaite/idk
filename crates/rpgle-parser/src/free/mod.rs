mod compiler_directives_parser;
mod dspec_keywords_parser;
mod free_parser;
mod fspec_keywords_parser;
mod hspec_keywords_parser;
mod op_parser;

pub mod core;
pub mod lexer;

pub use compiler_directives_parser::{tokenize_directive, DirectiveToken, DirectiveTokenKind};
pub use dspec_keywords_parser::{legacy_tokenize_dspec_kw, tokenize_dspec_kw, DToken, DTokenKind};
pub use free_parser::{
    legacy_tokenize, legacy_tokenize_extf2, tokenize, tokenize_extf2, tokenize_traditional_f2,
    Token, TokenKind,
};
pub use fspec_keywords_parser::{legacy_tokenize_fspec_kw, tokenize_fspec_kw, FToken, FTokenKind};
pub use hspec_keywords_parser::{legacy_tokenize_hspec_kw, tokenize_hspec_kw, HToken, HTokenKind};
pub use op_parser::Op;
