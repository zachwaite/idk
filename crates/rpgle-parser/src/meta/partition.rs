// based on: https://github.com/llogiq/partition/blob/master/src/lib.rs
// except I don't want mutable outputs
#[inline]
pub fn partition_index<T, P>(data: &mut [T], predicate: P) -> usize
where
    P: Fn(&T) -> bool,
{
    let len = data.len();
    if len == 0 {
        return 0;
    }
    let (mut l, mut r) = (0, len - 1);
    loop {
        while l < len && predicate(&data[l]) {
            l += 1;
        }
        while r > 0 && !predicate(&data[r]) {
            r -= 1;
        }
        if l >= r {
            return l;
        }
        data.swap(l, r);
    }
}

/* Slice partitioning. Split a slice into (keep, ignore) slices
 * using a predicate. Requires a mutable input slice as a trade off
 * to avoid the allocation from normal iter().partition(), which
 * allocates into vectors.
 */
pub fn partition<T, P>(data: &mut [T], predicate: P) -> (&[T], &[T])
where
    P: Fn(&T) -> bool,
{
    let idx = partition_index(data, predicate);
    return data.split_at(idx);
}
