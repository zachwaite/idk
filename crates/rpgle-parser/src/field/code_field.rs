use std::fmt::Display;

use super::result::{FieldBehavior, FieldResult};
use crate::free::Op;
use crate::meta::{Meta, Position, Span};
use nonempty::NonEmpty;
use serde::{Deserialize, Serialize};

// raw
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RawCodeField {
    pub value: NonEmpty<char>,
    pub meta: Meta,
}

impl Display for RawCodeField {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.meta.text)
    }
}
impl FieldBehavior for RawCodeField {
    fn span(&self) -> Span {
        self.meta.span
    }

    fn highlight(&self) -> Vec<(Span, String)> {
        vec![(self.span(), "Normal".to_string())]
    }
}
impl From<(Position, &[char; 93])> for FieldResult<RawCodeField> {
    fn from(value: (Position, &[char; 93])) -> Self {
        let pos = value.0;
        let chars = value.1;
        let meta = Meta::from((pos, chars.as_slice()));
        let value = NonEmpty::from_vec(chars.iter().map(|c| *c).collect::<Vec<char>>())
            .expect("&[char; 93] is guaranteed to be nonempty");
        Self::Ok(RawCodeField { value, meta })
    }
}

// cooked
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CodeField {
    pub op: Op,
}

impl Display for CodeField {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.op.to_string())
    }
}
impl FieldBehavior for CodeField {
    fn span(&self) -> Span {
        self.op.span()
    }

    fn highlight(&self) -> Vec<(Span, String)> {
        self.op.highlight()
    }
}
