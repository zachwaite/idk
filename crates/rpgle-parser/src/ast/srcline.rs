// Adapts SpecLine to SrcLine
use crate::field::{
    CommentField, CompilerDirectiveField, ControlLevelField, DatastructureTypeField, DatatypeField,
    DecimalsField, DefinitionTypeField, DeviceField, EndfileField, ExternalDescriptionField,
    Factor1Field, FieldResult, FileAdditionField, FileDesignationField, FileFormatField,
    FileOrganizationField, FileSequenceField, FiletypeField, FormtypeField, IdkField,
    IndicatorsField, KeyLengthField, LimitsProcessingField, NameField, NothingField,
    OperationField, POSField, RawCodeField, RawFactor2Field, RawKeywordsField,
    RecordAddressTypeField, RecordLengthField, ReservedField, ResultField, ResultLengthField,
    SequenceField,
};
use crate::line::{CSpecLine, SpecLine};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub enum CSrcline {
    Traditional {
        nothing: FieldResult<NothingField>,
        form_type: FieldResult<FormtypeField>,
        control_level: FieldResult<ControlLevelField>,
        indicators: FieldResult<IndicatorsField>,
        factor1: FieldResult<Factor1Field>,
        operation: FieldResult<OperationField>,
        factor2: FieldResult<RawFactor2Field>,
        result: FieldResult<ResultField>,
        result_length: FieldResult<ResultLengthField>,
        decimals: FieldResult<DecimalsField>,
        resulting_indicators: FieldResult<IndicatorsField>,
        comments: FieldResult<CommentField>,
    },
    ExtF2 {
        nothing: FieldResult<NothingField>,
        form_type: FieldResult<FormtypeField>,
        control_level: FieldResult<ControlLevelField>,
        indicators: FieldResult<IndicatorsField>,
        factor1: FieldResult<Factor1Field>,
        operation: FieldResult<OperationField>,
        factor2: FieldResult<RawFactor2Field>,
    },
    Free {
        nothing: FieldResult<NothingField>,
        code: FieldResult<RawCodeField>,
    },
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub enum Srcline {
    Idk {
        idk: FieldResult<IdkField>,
    },
    Comment {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        comment: FieldResult<CommentField>,
    },
    CompilerDirective {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        directive: FieldResult<CompilerDirectiveField>,
    },
    H {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        keywords: FieldResult<RawKeywordsField>,
    },
    F {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        name: FieldResult<NameField>,
        filetype: FieldResult<FiletypeField>,
        file_designation: FieldResult<FileDesignationField>,
        endfile: FieldResult<EndfileField>,
        file_addition: FieldResult<FileAdditionField>,
        file_sequence: FieldResult<FileSequenceField>,
        file_format: FieldResult<FileFormatField>,
        record_length: FieldResult<RecordLengthField>,
        limits_processing: FieldResult<LimitsProcessingField>,
        keylength: FieldResult<KeyLengthField>,
        record_address_type: FieldResult<RecordAddressTypeField>,
        file_organization: FieldResult<FileOrganizationField>,
        device: FieldResult<DeviceField>,
        reserved: FieldResult<ReservedField>,
        keywords: FieldResult<RawKeywordsField>,
    },
    FCont {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        nothing: FieldResult<NothingField>,
        keywords: FieldResult<RawKeywordsField>,
    },
    D {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        name: FieldResult<NameField>,
        external_description: FieldResult<ExternalDescriptionField>,
        datastructure_type: FieldResult<DatastructureTypeField>,
        definition_type: FieldResult<DefinitionTypeField>,
        from_position: FieldResult<POSField>,
        to_length: FieldResult<POSField>,
        datatype: FieldResult<DatatypeField>,
        decimals: FieldResult<DecimalsField>,
        reserved: FieldResult<ReservedField>,
        keywords: FieldResult<RawKeywordsField>,
    },
    DCont {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        nothing: FieldResult<NothingField>,
        keywords: FieldResult<RawKeywordsField>,
    },
    C(CSrcline),
}

pub fn srcline_from_specline(specline: &SpecLine) -> Srcline {
    match specline {
        SpecLine::Idk(line) => Srcline::Idk {
            idk: line.idk.clone(),
        },
        SpecLine::Comment(line) => Srcline::Comment {
            sequence: line.sequence.clone(),
            form_type: line.form_type.clone(),
            comment: line.comment.clone(),
        },
        SpecLine::CompilerDirective(line) => Srcline::CompilerDirective {
            sequence: line.sequence.clone(),
            form_type: line.form_type.clone(),
            directive: line.directive.clone(),
        },
        SpecLine::HSpec(line) => Srcline::H {
            sequence: line.sequence.clone(),
            form_type: line.form_type.clone(),
            keywords: line.keywords.clone(),
        },
        SpecLine::FSpec(line) => Srcline::F {
            sequence: line.sequence.clone(),
            form_type: line.form_type.clone(),
            name: line.name.clone(),
            filetype: line.filetype.clone(),
            file_designation: line.file_designation.clone(),
            endfile: line.endfile.clone(),
            file_addition: line.file_addition.clone(),
            file_sequence: line.file_sequence.clone(),
            file_format: line.file_format.clone(),
            record_length: line.record_length.clone(),
            limits_processing: line.limits_processing.clone(),
            keylength: line.keylength.clone(),
            record_address_type: line.record_address_type.clone(),
            file_organization: line.file_organization.clone(),
            device: line.device.clone(),
            reserved: line.reserved.clone(),
            keywords: line.keywords.clone(),
        },
        SpecLine::FSpecContinuation(line) => Srcline::FCont {
            sequence: line.sequence.clone(),
            form_type: line.form_type.clone(),
            nothing: line.nothing.clone(),
            keywords: line.keywords.clone(),
        },
        SpecLine::DSpec(line) => Srcline::D {
            sequence: line.sequence.clone(),
            form_type: line.form_type.clone(),
            name: line.name.clone(),
            external_description: line.external_description.clone(),
            datastructure_type: line.datastructure_type.clone(),
            definition_type: line.definition_type.clone(),
            from_position: line.from_position.clone(),
            to_length: line.to_length.clone(),
            datatype: line.datatype.clone(),
            decimals: line.decimals.clone(),
            reserved: line.reserved.clone(),
            keywords: line.keywords.clone(),
        },
        SpecLine::DSpecContinuation(line) => Srcline::DCont {
            sequence: line.sequence.clone(),
            form_type: line.form_type.clone(),
            nothing: line.nothing.clone(),
            keywords: line.keywords.clone(),
        },
        SpecLine::CSpec(line) => match line {
            CSpecLine::Traditional(line) => Srcline::C(CSrcline::Traditional {
                nothing: line.nothing.clone(),
                form_type: line.form_type.clone(),
                control_level: line.control_level.clone(),
                indicators: line.indicators.clone(),
                factor1: line.factor1.clone(),
                operation: line.operation.clone(),
                factor2: line.factor2.clone(),
                result: line.result.clone(),
                result_length: line.result_length.clone(),
                decimals: line.decimals.clone(),
                resulting_indicators: line.resulting_indicators.clone(),
                comments: line.comments.clone(),
            }),
            CSpecLine::ExtF2(line) => Srcline::C(CSrcline::ExtF2 {
                nothing: line.nothing.clone(),
                form_type: line.form_type.clone(),
                control_level: line.control_level.clone(),
                indicators: line.indicators.clone(),
                factor1: line.factor1.clone(),
                operation: line.operation.clone(),
                factor2: line.factor2.clone(),
            }),
            CSpecLine::Free(line) => Srcline::C(CSrcline::Free {
                nothing: line.nothing.clone(),
                code: line.code.clone(),
            }),
        },
    }
}
