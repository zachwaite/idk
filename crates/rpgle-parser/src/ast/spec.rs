use super::srcline::{CSrcline, Srcline};
use crate::field::{
    CodeField, DKeywordsField, DatastructureTypeField, DatatypeField, DecimalsField,
    DefinitionTypeField, DeviceField, EndfileField, ExternalDescriptionField, FKeywordsField,
    FieldResult, FileAdditionField, FileDesignationField, FileFormatField, FileOrganizationField,
    FileSequenceField, FiletypeField, FormtypeField, HKeywordsField, KeyLengthField,
    LimitsProcessingField, NameField, POSField, RecordAddressTypeField, RecordLengthField,
    ReservedField, SequenceField,
};

use crate::free::{
    legacy_tokenize_dspec_kw, legacy_tokenize_fspec_kw, legacy_tokenize_hspec_kw, Op,
};
use crate::meta::partition::partition;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum Spec {
    H {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        keywords: FieldResult<HKeywordsField>,
    },
    F {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        name: FieldResult<NameField>,
        filetype: FieldResult<FiletypeField>,
        file_designation: FieldResult<FileDesignationField>,
        endfile: FieldResult<EndfileField>,
        file_addition: FieldResult<FileAdditionField>,
        file_sequence: FieldResult<FileSequenceField>,
        file_format: FieldResult<FileFormatField>,
        record_length: FieldResult<RecordLengthField>,
        limits_processing: FieldResult<LimitsProcessingField>,
        keylength: FieldResult<KeyLengthField>,
        record_address_type: FieldResult<RecordAddressTypeField>,
        file_organization: FieldResult<FileOrganizationField>,
        device: FieldResult<DeviceField>,
        reserved: FieldResult<ReservedField>,
        keywords: FieldResult<FKeywordsField>,
    },
    D {
        sequence: FieldResult<SequenceField>,
        form_type: FieldResult<FormtypeField>,
        name: FieldResult<NameField>,
        external_description: FieldResult<ExternalDescriptionField>,
        datastructure_type: FieldResult<DatastructureTypeField>,
        definition_type: FieldResult<DefinitionTypeField>,
        from_position: FieldResult<POSField>,
        to_length: FieldResult<POSField>,
        datatype: FieldResult<DatatypeField>,
        decimals: FieldResult<DecimalsField>,
        reserved: FieldResult<ReservedField>,
        keywords: FieldResult<DKeywordsField>,
    },
    C {
        code: FieldResult<CodeField>,
    },
}

#[derive(Debug)]
pub enum ParseError {
    EmptyInput,
    Unhandled,
}

// H ---------------------------------------------------------------------------------
fn try_hspec(input: &[Srcline]) -> Option<(Spec, &[Srcline])> {
    let first = input.get(0)?;
    let spec = match first {
        Srcline::H {
            sequence,
            form_type,
            keywords,
        } => {
            let tokens = legacy_tokenize_hspec_kw(keywords);
            let spec = Spec::H {
                sequence: sequence.clone(),
                form_type: form_type.clone(),
                keywords: FieldResult::Ok(HKeywordsField { tokens }),
            };
            Some(spec)
        }
        Srcline::Idk { .. }
        | Srcline::Comment { .. }
        | Srcline::CompilerDirective { .. }
        | Srcline::F { .. }
        | Srcline::FCont { .. }
        | Srcline::D { .. }
        | Srcline::DCont { .. }
        | Srcline::C(_) => None,
    }?;
    if input.len() > 1 {
        Some((spec, &input[1..]))
    } else {
        Some((spec, &[]))
    }
}

// YAGNI
// fn hspec(input: &[Srcline]) -> Result<(Spec, &[Srcline]), ParseError> {
//     try_hspec(&input).ok_or(ParseError::Unhandled)
// }

// F ---------------------------------------------------------------------------------
fn try_fspec(input: &[Srcline]) -> Option<(Spec, &[Srcline])> {
    let mut idx = 0;
    let first = input.get(idx)?;
    idx += 1;
    let spec = match first {
        Srcline::F {
            sequence,
            form_type,
            name,
            filetype,
            file_designation,
            endfile,
            file_addition,
            file_sequence,
            file_format,
            record_length,
            limits_processing,
            keylength,
            record_address_type,
            file_organization,
            device,
            reserved,
            keywords,
        } => {
            let mut cont_keywords = vec![];
            // collect kw results from continuations
            loop {
                if let Some(Srcline::FCont { keywords: kw, .. }) = input.get(idx) {
                    cont_keywords.push(kw);
                    idx += 1;
                } else {
                    break;
                }
            }
            // parse fspec kw into ftokens
            let tokens = legacy_tokenize_fspec_kw(keywords, &cont_keywords);
            let spec = Spec::F {
                sequence: sequence.clone(),
                form_type: form_type.clone(),
                name: name.clone(),
                filetype: filetype.clone(),
                file_designation: file_designation.clone(),
                endfile: endfile.clone(),
                file_addition: file_addition.clone(),
                file_sequence: file_sequence.clone(),
                file_format: file_format.clone(),
                record_length: record_length.clone(),
                limits_processing: limits_processing.clone(),
                keylength: keylength.clone(),
                record_address_type: record_address_type.clone(),
                file_organization: file_organization.clone(),
                device: device.clone(),
                reserved: reserved.clone(),
                keywords: FieldResult::Ok(FKeywordsField { tokens }),
            };
            Some(spec)
        }
        Srcline::Idk { .. }
        | Srcline::Comment { .. }
        | Srcline::CompilerDirective { .. }
        | Srcline::H { .. }
        | Srcline::FCont { .. }
        | Srcline::D { .. }
        | Srcline::DCont { .. }
        | Srcline::C(_) => None,
    }?;
    if input.len() > 1 {
        Some((spec, &input[idx..]))
    } else {
        Some((spec, &[]))
    }
}

// D ---------------------------------------------------------------------------------
fn try_dspec(input: &[Srcline]) -> Option<(Spec, &[Srcline])> {
    let mut idx = 0;
    let first = input.get(idx)?;
    idx += 1;
    let spec = match first {
        Srcline::D {
            sequence,
            form_type,
            name,
            external_description,
            datastructure_type,
            definition_type,
            from_position,
            to_length,
            datatype,
            decimals,
            reserved,
            keywords,
        } => {
            let mut cont_keywords = vec![];
            // collect continuations
            loop {
                if let Some(Srcline::DCont { keywords: kw, .. }) = input.get(idx) {
                    cont_keywords.push(kw);
                    idx += 1;
                } else {
                    break;
                }
            }
            // parse dspec kw into dtokens
            let tokens = legacy_tokenize_dspec_kw(keywords, &cont_keywords);
            let spec = Spec::D {
                sequence: sequence.clone(),
                form_type: form_type.clone(),
                name: name.clone(),
                external_description: external_description.clone(),
                datastructure_type: datastructure_type.clone(),
                definition_type: definition_type.clone(),
                from_position: from_position.clone(),
                to_length: to_length.clone(),
                datatype: datatype.clone(),
                decimals: decimals.clone(),
                reserved: reserved.clone(),
                keywords: FieldResult::Ok(DKeywordsField { tokens }),
            };
            Some(spec)
        }
        Srcline::Idk { .. }
        | Srcline::Comment { .. }
        | Srcline::CompilerDirective { .. }
        | Srcline::H { .. }
        | Srcline::F { .. }
        | Srcline::FCont { .. }
        | Srcline::DCont { .. }
        | Srcline::C(_) => None,
    }?;
    if input.len() > 1 {
        Some((spec, &input[idx..]))
    } else {
        Some((spec, &[]))
    }
}

// C ---------------------------------------------------------------------------------
fn try_cspec_free(input: &[Srcline]) -> Option<(Spec, &[Srcline])> {
    let mut idx = 0;
    let first = input.get(0)?;
    idx += 1;
    let spec = match first {
        Srcline::C(CSrcline::Free {
            code: codefield, ..
        }) => {
            // TODO: collect continuations
            let conts = vec![];
            // TODO: Op parser uses legacy style parser
            let op = Op::from((codefield, conts.as_slice()));
            let fld = CodeField { op };
            let code = FieldResult::Ok(fld);
            let spec = Spec::C { code };
            Some(spec)
        }
        Srcline::Idk { .. }
        | Srcline::Comment { .. }
        | Srcline::CompilerDirective { .. }
        | Srcline::H { .. }
        | Srcline::F { .. }
        | Srcline::FCont { .. }
        | Srcline::D { .. }
        | Srcline::DCont { .. }
        | Srcline::C(_) => None,
    }?;
    if input.len() > 1 {
        Some((spec, &input[idx..]))
    } else {
        Some((spec, &[]))
    }
}

fn try_cspec_extf2(input: &[Srcline]) -> Option<(Spec, &[Srcline])> {
    todo!()
}

fn try_cspec_traditional(input: &[Srcline]) -> Option<(Spec, &[Srcline])> {
    let mut idx = 0;
    let first = input.get(0)?;
    idx += 1;
    let spec = match first {
        Srcline::C(CSrcline::Traditional {
            factor1, operation, ..
        }) => {
            // TODO: collect continuations
            let conts = vec![];
            // TODO: Op parser uses legacy style parser
            let op = Op::from(((operation, factor1), conts.as_slice()));
            let fld = CodeField { op };
            let code = FieldResult::Ok(fld);
            let spec = Spec::C { code };
            Some(spec)
        }
        Srcline::Idk { .. }
        | Srcline::Comment { .. }
        | Srcline::CompilerDirective { .. }
        | Srcline::H { .. }
        | Srcline::F { .. }
        | Srcline::FCont { .. }
        | Srcline::D { .. }
        | Srcline::DCont { .. }
        | Srcline::C(_) => None,
    }?;
    if input.len() > 1 {
        Some((spec, &input[idx..]))
    } else {
        Some((spec, &[]))
    }
}

fn spec(input: &[Srcline]) -> Result<(Spec, &[Srcline]), ParseError> {
    // This parser implements the `choice` pattern using partial parsers.
    // Input flows through the first parser and if the parser succeeds,
    // it will return, else input flows to the next parser.
    // If no parsers succeed, a ParseError is returned.
    let parse_hspec = || try_hspec(input);
    let parse_fspec = || try_fspec(input);
    let parse_dspec = || try_dspec(input);
    let parse_cspec_free = || try_cspec_free(input);
    // let parse_cspec_extf2 = || try_cspec_extf2(input); // Not Implemented
    let parse_cspec_traditional = || try_cspec_traditional(input);
    parse_hspec()
        .or_else(parse_fspec)
        .or_else(parse_dspec)
        .or_else(parse_cspec_free)
        .or_else(parse_cspec_traditional)
        .ok_or(ParseError::Unhandled)
}

pub fn ast(input: &mut [Srcline]) -> Result<(Vec<Spec>, &[Srcline]), ParseError> {
    // This filter splits the input into things you can handle and things you
    // can't. This will prevent erroring out due to not implemented features.
    // This is mostly useful for development, but think of it like a breaker
    // box or water supply manifold... it's better to have it available for
    // maintenance than to need to stop using the whole thing.
    let (keep, ignore) = partition(input, |line| match line {
        Srcline::Idk { .. } => false,
        Srcline::Comment { .. } => false,
        Srcline::CompilerDirective { .. } => false,
        Srcline::H { .. } => true,
        Srcline::F { .. } => true,
        Srcline::FCont { .. } => true,
        Srcline::D { .. } => true,
        Srcline::DCont { .. } => true,
        Srcline::C(_) => true,
    });
    let mut _input = keep;
    let mut outs: Vec<Spec> = vec![];
    loop {
        let (s, _rest) = spec(&_input)?;
        outs.push(s);
        if _rest.len() == 0 {
            break;
        } else {
            _input = _rest;
        }
    }
    Ok((outs, ignore))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{ast::srcline::srcline_from_specline, cst::CST};
    use insta;

    #[test]
    fn test_hspec_01() {
        // empty input
        let lines = vec![];
        let observed = try_hspec(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_hspec_02() {
        // 1 hline
        let input = r#"
     H OPTION(*nodebugio:*srcstmt)                                                                  "#[1..].to_string();
        let cst = CST::try_from(input.as_str()).unwrap();
        let lines = cst
            .lines
            .iter()
            .map(|line| srcline_from_specline(line))
            .collect::<Vec<Srcline>>();
        let observed = try_hspec(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_fspec_01() {
        // empty input
        let lines = vec![];
        let observed = try_fspec(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_fspec_02() {
        // 1 fline
        let input = r#"
     FCowEvt    UF A E           K DISK                                                             "#
            [1..].to_string();
        let cst = CST::try_from(input.as_str()).unwrap();
        let lines = cst
            .lines
            .iter()
            .map(|line| srcline_from_specline(line))
            .collect::<Vec<Srcline>>();
        let observed = try_fspec(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_fspec_03() {
        // fline + continuation
        let input = r#"
     FCowEvtL2  IF   E           K DISK     Rename(EVTFMT:VEVTFMT)                                  
     F                                     Prefix(V)                                                "#
            [1..].to_string();
        let cst = CST::try_from(input.as_str()).unwrap();
        let lines = cst
            .lines
            .iter()
            .map(|line| srcline_from_specline(line))
            .collect::<Vec<Srcline>>();
        let observed = try_fspec(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_dspec_01() {
        // empty input
        let lines = vec![];
        let observed = try_dspec(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_dspec_02() {
        // 1 dline
        let input = r#"
     D LastId          S              8  0                                                          "#
            [1..].to_string();
        let cst = CST::try_from(input.as_str()).unwrap();
        let lines = cst
            .lines
            .iter()
            .map(|line| srcline_from_specline(line))
            .collect::<Vec<Srcline>>();
        let observed = try_dspec(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_cspec_free_01() {
        // empty input
        let lines = vec![];
        let observed = try_cspec_free(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_cspec_free_02() {
        // 1 cline
        let input = r#"
       Exsr $SetLstId;                                                                              "#
            [1..].to_string();
        let cst = CST::try_from(input.as_str()).unwrap();
        let lines = cst
            .lines
            .iter()
            .map(|line| srcline_from_specline(line))
            .collect::<Vec<Srcline>>();
        let observed = try_cspec_free(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    // TODO: extf2

    #[test]
    fn test_cspec_traditional_01() {
        // empty input
        let lines = vec![];
        let observed = try_cspec_traditional(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_cspec_traditional_02() {
        // 1 cline
        let input = r#"
     C     $CrtBRNEVT    BegSr                                                                      "#
            [1..].to_string();
        let cst = CST::try_from(input.as_str()).unwrap();
        let lines = cst
            .lines
            .iter()
            .map(|line| srcline_from_specline(line))
            .collect::<Vec<Srcline>>();
        let observed = try_cspec_traditional(&lines);
        insta::assert_yaml_snapshot!(observed);
    }

    #[test]
    fn test_ast_snapshot() {
        let input = &r#"
     H OPTION(*nodebugio:*srcstmt)                                                                  
     FCowEvt    UF A E           K DISK                                                             
     FBornEvt   UF A E           K DISK                                                             
     FCowEvtL2  IF   E           K DISK     Rename(EVTFMT:VEVTFMT)                                  
     F                                     Prefix(V)                                                
     F**********************************************************************************************
     D**********************************************************************************************
     D LastId          S              8  0                                                          
     D QCmdExc         PR                  EXTPGM('QCMDEXC')                                        
     D  Command                    2000                                                             
     D  Length                       15  5                                                          
     C**********************************************************************************************
      /free                                                                                         
       Exsr $SetLstId;                                                                              
       Exsr $CrtEvts;                                                                               
       QCmdExc(Foo:Bar);                                                                            
       *inlr = *on;                                                                                 
                                                                                                    
       Begsr $SetLstId;                                                                             
         SetLL *Loval CowEvtL2;                                                                     
         If Not %Eof;                                                                               
           Read CowEvtL2;                                                                           
             QCmdExc(FOO:BaR);                                                                      
           LastId = Vid;                                                                            
         Else;                                                                                      
          LastId = 1;                                                                               
         Endif;                                                                                     
       Endsr;                                                                                       
                                                                                                    
     C     $CrtBRNEVT    BegSr                                                                      
         EID = Id;                                                                                  
         BNAME = 'BESSE';                                                                           
         BDAT = 20240101;                                                                           
         Write BORNFMT;                                                                             
     C                   ENDSR                                                                      
                                                                                                    
       Begsr $CrtCowEvt;                                                                            
         Id = LastId + 1;                                                                           
         Edat = 20240101;                                                                           
         Etim = 125959;                                                                             
         Etyp = 'BORN';                                                                             
         Write EVTFMT;                                                                              
       Endsr;                                                                                       
                                                                                                    
       Begsr $CrtEvts;                                                                              
         Exsr $CrtCowEvt;                                                                           
         Exsr $CrtBrnEvt;                                                                           
       Endsr;                                                                                       "#
            [1..];
        let cst = CST::try_from(input).unwrap();
        let mut lines = cst
            .lines
            .iter()
            .map(|line| srcline_from_specline(line))
            .collect::<Vec<Srcline>>();
        let (specs, rest) = ast(&mut lines).unwrap();
        for ln in rest.iter() {
            println!("{:?}", ln);
        }
        // expect 3 Comments and a CompilerDirective to be ignored
        assert_eq!(rest.len(), 4);
        insta::assert_yaml_snapshot!(specs);
    }
}
