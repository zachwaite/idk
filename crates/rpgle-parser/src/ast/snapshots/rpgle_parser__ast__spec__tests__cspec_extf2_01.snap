---
source: crates/rpgle-parser/src/ast/spec.rs
assertion_line: 470
expression: observed
---
~
